prover/src/lib.rs

```diff
@@ -211,39 +300,204 @@ impl SP1Prover {
+                SP1ReduceProofWrapper::Recursive(ref proof) => {
+                    ReduceState::from_reduce_start_state(proof)
+                }
+            })
+            .collect::<Vec<_>>();
+        for start_state in start_states.iter() {
+            println!(
+                "1deferred_digest = {:?}",
+                start_state.deferred_proofs_digest
+            );
+            println!(
+                "1deferred_digest = {:?}",
+                start_state.reconstruct_deferred_digest
+            );
+        }
+        // This is the last layer only if the outcome is a single proof. If there are deferred proofs
+        // or there is a single proof being pushed to the next layer, it's not the last layer.
+        let is_complete = chunks.len() == 1 && last_proof.is_none() && deferred_proofs.is_empty();
+        println!(
+            "num main chunks = {}, num deferred chunks = {}",
+            chunks.len(),
+            deferred_proofs.len()
+        );
         let mut new_proofs: Vec<SP1ReduceProofWrapper> = chunks
-            .into_par_iter()
-            .map(|chunk| {
-                let proof = self.reduce_batch(vk, sp1_challenger.clone(), chunk, &[]);
+            .into_iter()
+            .zip(reconstruct_challengers.into_iter())
+            .zip(start_states.into_iter())
+            .map(|((chunk, reconstruct_challenger), start_state)| {
+                let proof = self.reduce_batch(
+                    vk,
+                    sp1_challenger.clone(),
+                    reconstruct_challenger,
+                    start_state,
+                    chunk,
+                    &[],
+                    is_complete,
+                );
                 SP1ReduceProofWrapper::Recursive(proof)
             })
             .collect();
 
         if let Some(proof) = last_proof {
             new_proofs.push(proof);
         }
+
+        // For all the proofs with only deferred proofs, the start and end state will be the end
+        // state of the last proof from above.
+        let last_new_proof = &new_proofs[new_proofs.len() - 1];
+        let mut reduce_state: ReduceState = match last_new_proof {
+            SP1ReduceProofWrapper::Recursive(ref proof) => {
+                ReduceState::from_reduce_end_state(proof)
+            }
+            _ => unreachable!(),
+        };
+        let deferred_chunks: Vec<_> = deferred_proofs.chunks(batch_size).collect();
+        let start_states = deferred_chunks
+            .iter()
+            .map(|chunk| {
+                let start_state = reduce_state.clone();
+                // Accumulate deferred proofs into the digest
+                // poseidon2( current_digest[..8] || pv.sp1_vk_digest[..8] || pv.committed_value_digest[..32] )
+                for proof in chunk.iter() {
+                    println!(
+                        "before deferred_digest = {:?}",
+                        reduce_state.reconstruct_deferred_digest
+                    );
```

Comment:
somethng about this also seems wrong, we should try to avoid manual serialization inside the prover/src/lib.rs file